crates/ruff/src/rules/flake8_simplify/rules/ast_with.rs

```diff
@@ -60,9 +60,15 @@ impl Violation for MultipleWithStatements {
     }
 }
 
-fn find_last_with(body: &[Stmt]) -> Option<(&[Withitem], &[Stmt])> {
-    let [Stmt::With(ast::StmtWith { items, body, .. })] = body else { return None };
-    find_last_with(body).or(Some((items, body)))
+/// Returns a boolean indicating its an async with statement, the items and
+/// body.
+fn find_last_with(body: &[Stmt]) -> Option<(bool, &[Withitem], &[Stmt])> {
+    let (is_async, items, body) = match body {
+        [Stmt::With(ast::StmtWith { items, body, .. })] => (false, items, body),
+        [Stmt::AsyncWith(ast::StmtAsyncWith { items, body, .. })] => (true, items, body),
+        _ => return None,
+    };
+    find_last_with(body).or(Some((is_async, items, body)))
```

Comment:
What's the purpose of the recursion if we only fix one level in each ruff fix iteration?

Reply:
I'm not exactly sure, I didn't want to change the existing code too much possibly breaking something in the process.

It does slightly improve the suggestion span (`-` is old code, `+` is with the recursive call):

```
---- rules::flake8_simplify::tests::rules::rule_multiplewithstatements_path_new_sim117_py_expects stdout ----
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Snapshot Summary ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Snapshot file: crates/ruff/src/rules/flake8_simplify/snapshots/ruff__rules__flake8_simplify__tests__SIM117_SIM117.py.snap
Snapshot: SIM117_SIM117.py
Source: crates/ruff/src/rules/flake8_simplify/mod.rs:52
────────────────────────────────────────────────────────────────────────────────────────────────────────
-old snapshot
+new results
────────────┬───────────────────────────────────────────────────────────────────────────────────────────
   22    22 │    |
   23    23 │  6 |   # SIM117
   24    24 │  7 | / with A():
   25    25 │  8 | |     with B():
   26       │- 9 | |         with C():
   27       │-   | |_________________^ SIM117
         26 │+   | |_____________^ SIM117
         27 │+ 9 |           with C():
   28    28 │ 10 |               print("hello")
   29    29 │    |
   30    30 │    = help: Combine `with` statements
   31    31 │
────────────┴───────────────────────────────────────────────────────────────────────────────────────────

```

But perhaps we can resolve that without recursion.

Reply:
> What's the purpose of the recursion if we only fix one level in each ruff fix iteration?

Is this true? I was under the impression that we did collapse multiple levels (haven't looked back at the code).

Reply:
After following some red hearings I think I can say that we don't completely collapse multiple levels. This is the output of the current code (before this pr):

```
SIM117.py:2:1: SIM117 [*] Use a single `with` statement with multiple contexts instead of nested `with` statements
  |
1 |   # SIM117
2 | / with A():
3 | |     with B():
4 | |         with C():
  | |_________________^ SIM117
5 |               print("hello")
  |
  = help: Combine `with` statements

ℹ Suggested fix
1 1 | # SIM117
2   |-with A():
3   |-    with B():
4   |-        with C():
5   |-            print("hello")
  2 |+with A(), B():
  3 |+    with C():
  4 |+        print("hello")
```

Note that the error indicator (first part) includes all three `with` statements, but the suggested fix only includes the first two `with` statement. Which is not ideal because when applied Ruff will complain the next run to also collapse `with C()`.

I could work on the fix suggestion to rewrite it to `with A(), B(), C()`, but that will have a larger scope than issue https://github.com/astral-sh/ruff/issues/3025 is aiming to fix, so perhaps we should do it in another pr.

Reply:
Ahh ok, so we flag them all, but only fix one level... This is not _terrible_ because Ruff runs iteratively until there are no more fixable errors, so from the user's perspective, it will appear as if these were fixed in one pass. But it does mean that the suggested fix is confusing as depicted above.

Let's fix it if possible, but in a separate PR.